tedana/model/monoexponential.py

```diff
@@ -133,58 +155,23 @@ def fit_decay_ts(data, mask, tes, masksum, start_echo):
     s0vaf : (S x E x T) :obj:`numpy.ndarray`
         Full S0 map
     """
-    nx, ny, nz, n_echos, n_trs = data.shape
+    n_samples, _, n_trs = data.shape
     echodata = data[mask]
     tes = np.array(tes)
 
-    t2sa_ts = np.zeros([nx, ny, nz, n_trs])
-    s0va_ts = np.zeros([nx, ny, nz, n_trs])
-    t2saf_ts = np.zeros([nx, ny, nz, n_trs])
-    s0vaf_ts = np.zeros([nx, ny, nz, n_trs])
+    t2sa_ts = np.zeros([n_samples, n_trs])
+    s0va_ts = np.copy(t2sa_ts)
+    t2saf_ts = np.copy(t2sa_ts)
+    s0vaf_ts = np.copy(t2sa_ts)
 
     for vol in range(echodata.shape[-1]):
-        t2ss = np.zeros([nx, ny, nz, n_echos - 1])
-        s0vs = t2ss.copy()
-        # Fit monoexponential decay first for first echo only,
-        # then first two echoes, etc.
-        for i_echo in range(start_echo, n_echos + 1):
-            B = np.abs(echodata[:, :i_echo, vol]) + 1
-            B = np.log(B).transpose()
-            neg_tes = -1 * tes[:i_echo]
-
-            # First row is constant, second is TEs for decay curve
-            # Independent variables for least-squares model
-            x = np.array([np.ones(i_echo), neg_tes])
-            X = np.sort(x)[:, ::-1].transpose()
-
-            beta, _, _, _ = np.linalg.lstsq(X, B)
-            t2s = 1. / beta[1, :].transpose()
-            s0 = np.exp(beta[0, :]).transpose()
-
-            t2s[np.isinf(t2s)] = 500.
-            s0[np.isnan(s0)] = 0.
-
-            t2ss[:, :, :, i_echo-2] = np.squeeze(utils.unmask(t2s, mask))
-            s0vs[:, :, :, i_echo-2] = np.squeeze(utils.unmask(s0, mask))
-
-        # Limited T2* and S0 maps
-        fl = np.zeros([nx, ny, nz, len(tes)-1], bool)
-        for i_echo in range(n_echos - 1):
-            fl_ = np.squeeze(fl[:, :, :, i_echo])
-            fl_[masksum == i_echo + 2] = True
-            fl[:, :, :, i_echo] = fl_
-        t2sa = np.squeeze(utils.unmask(t2ss[fl], masksum > 1))
-        s0va = np.squeeze(utils.unmask(s0vs[fl], masksum > 1))
-
-        # Full T2* maps with S0 estimation errors
-        t2saf = t2sa.copy()
-        s0vaf = s0va.copy()
-        t2saf[masksum == 1] = t2ss[masksum == 1, 0]
-        s0vaf[masksum == 1] = s0vs[masksum == 1, 0]
-
-        t2sa_ts[:, :, :, vol] = t2sa
-        s0va_ts[:, :, :, vol] = s0va
-        t2saf_ts[:, :, :, vol] = t2saf
-        s0vaf_ts[:, :, :, vol] = s0vaf
+
+        [t2sa, s0va, _, _, t2saf, s0vaf] = fit_decay(
```

Comment:
What do you think about converting this to, instead of calling `fit_decay` inside of `fit_decay_ts`, being able to throw a `timeseries=True` flag ?

Reply:
@TomMaullin if you'd like to address this in a different PR, I can merge this one and leave #25 open !

Reply:
Personally, I think it would be good to leave that to a separate PR! This ones looks good to me :smile:

Reply:
Hi @emdupre , @rmarkello 

Sorry for the delayed response! Would you still like me to address this?